File: src/lib/logger.ts
// 日志模块 - 支持日志级别控制

export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
  NONE = 4
}

// 从环境变量读取日志级别，默认 INFO
const LOG_LEVEL_MAP: Record<string, LogLevel> = {
  'debug': LogLevel.DEBUG,
  'info': LogLevel.INFO,
  'warn': LogLevel.WARN,
  'error': LogLevel.ERROR,
  'none': LogLevel.NONE
}

function getLogLevel(): LogLevel {
  const level = process.env.LOG_LEVEL?.toLowerCase() || 'info'
  return LOG_LEVEL_MAP[level] ?? LogLevel.INFO
}

// 格式化时间戳
function getTimestamp(): string {
  return new Date().toISOString()
}

// 格式化日志前缀
function formatPrefix(level: string, module?: string): string {
  const timestamp = getTimestamp()
  const moduleStr = module ? `[${module}]` : ''
  return `${timestamp} ${level} ${moduleStr}`
}

// 日志工具
export const logger = {
  debug(message: string, data?: any, module?: string) {
    if (getLogLevel() <= LogLevel.DEBUG) {
      const prefix = formatPrefix('DEBUG', module)
      if (data !== undefined) {
        console.log(`${prefix} ${message}`, typeof data === 'object' ? JSON.stringify(data, null, 2) : data)
      } else {
        console.log(`${prefix} ${message}`)
      }
    }
  },

  info(message: string, data?: any, module?: string) {
    if (getLogLevel() <= LogLevel.INFO) {
      const prefix = formatPrefix('INFO ', module)
      if (data !== undefined) {
        console.log(`${prefix} ${message}`, typeof data === 'object' ? JSON.stringify(data, null, 2) : data)
      } else {
        console.log(`${prefix} ${message}`)
      }
    }
  },

  warn(message: string, data?: any, module?: string) {
    if (getLogLevel() <= LogLevel.WARN) {
      const prefix = formatPrefix('WARN ', module)
      if (data !== undefined) {
        console.warn(`${prefix} ${message}`, typeof data === 'object' ? JSON.stringify(data, null, 2) : data)
      } else {
        console.warn(`${prefix} ${message}`)
      }
    }
  },

  error(message: string, error?: any, module?: string) {
    if (getLogLevel() <= LogLevel.ERROR) {
      const prefix = formatPrefix('ERROR', module)
      if (error instanceof Error) {
        console.error(`${prefix} ${message}`, {
          message: error.message,
          stack: error.stack
        })
      } else if (error !== undefined) {
        console.error(`${prefix} ${message}`, typeof error === 'object' ? JSON.stringify(error, null, 2) : error)
      } else {
        console.error(`${prefix} ${message}`)
      }
    }
  },

  // 创建模块化日志器
  module(name: string) {
    return {
      debug: (message: string, data?: any) => logger.debug(message, data, name),
      info: (message: string, data?: any) => logger.info(message, data, name),
      warn: (message: string, data?: any) => logger.warn(message, data, name),
      error: (message: string, error?: any) => logger.error(message, error, name),
    }
  }
}

export default logger

File: backend/app/utils/logger.py
import json
import logging
import sys
import time
from datetime import datetime, timezone
from ..config import settings

# 日志级别映射
LOG_LEVEL_MAP = {
    "debug": logging.DEBUG,
    "info": logging.INFO,
    "warn": logging.WARNING,
    "error": logging.ERROR,
    "none": logging.CRITICAL,
}


def get_log_level() -> int:
    level = settings.LOG_LEVEL.lower()
    return LOG_LEVEL_MAP.get(level, logging.INFO)


class JsonFormatter(logging.Formatter):
    """JSON 结构化日志格式。"""

    def format(self, record):
        log_entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": record.levelname.lower(),
            "logger": record.name,
            "message": record.getMessage(),
        }
        if record.exc_info:
            log_entry["exception"] = self.formatException(record.exc_info)
        # 合并 extra 字段
        for key in ("request_id", "method", "path", "status_code", "duration_ms", "user_id"):
            val = getattr(record, key, None)
            if val is not None:
                log_entry[key] = val
        return json.dumps(log_entry, ensure_ascii=False)


class TextFormatter(logging.Formatter):
    """人类可读日志格式。"""

    def format(self, record):
        parts = [self.formatTime(record), record.levelname, f"[{record.name}]"]
        parts.append(record.getMessage())
        extras = {}
        for key in ("request_id", "method", "path", "status_code", "duration_ms"):
            val = getattr(record, key, None)
            if val is not None:
                extras[key] = val
        if extras:
            parts.append(json.dumps(extras, ensure_ascii=False))
        if record.exc_info:
            parts.append("\n" + self.formatException(record.exc_info))
        return " ".join(parts)


# 配置根日志记录器
_use_json = settings.LOG_LEVEL == "json"
_handler = logging.StreamHandler(sys.stdout)
_handler.setFormatter(JsonFormatter() if _use_json else TextFormatter())

root_logger = logging.getLogger()
root_logger.setLevel(get_log_level() if not _use_json else logging.INFO)
root_logger.handlers = [_handler]


class ModuleLogger:
    def __init__(self, module_name: str):
        self.logger = logging.getLogger(module_name)

    def debug(self, message: str, data=None):
        self.logger.debug(f"{message} {data}" if data else message)

    def info(self, message: str, data=None):
        self.logger.info(f"{message} {data}" if data else message)

    def warn(self, message: str, data=None):
        self.logger.warning(f"{message} {data}" if data else message)

    def error(self, message: str, error=None):
        if error and isinstance(error, Exception):
            self.logger.error(f"{message} {error}", exc_info=True)
        elif error:
            self.logger.error(f"{message} {error}")
        else:
            self.logger.error(message)


def create_module_logger(module_name: str) -> ModuleLogger:
    return ModuleLogger(module_name)
